questionpy_sdk/watcher.py

```diff
@@ -0,0 +1,196 @@
+#  The QuestionPy SDK is free software released under terms of the MIT license. See LICENSE.md.
+#  (c) Technische Universität Berlin, innoCampus <info@isis.tu-berlin.de>
+
+import asyncio
+import logging
+import sys
+from collections.abc import Awaitable, Callable
+from contextlib import AbstractAsyncContextManager
+from functools import cached_property
+from pathlib import Path
+from types import TracebackType
+from typing import TYPE_CHECKING, Self
+
+from watchdog.events import (
+    FileClosedEvent,
+    FileOpenedEvent,
+    FileSystemEvent,
+    FileSystemEventHandler,
+    FileSystemMovedEvent,
+)
+from watchdog.observers import Observer
+from watchdog.utils.event_debouncer import EventDebouncer
+
+from questionpy_common.constants import DIST_DIR
+from questionpy_sdk.package.builder import DirPackageBuilder
+from questionpy_sdk.package.errors import PackageBuildError, PackageSourceValidationError
+from questionpy_sdk.package.source import PackageSource
+from questionpy_sdk.webserver.app import WebServer
+from questionpy_server.worker.runtime.package_location import DirPackageLocation
+
+if TYPE_CHECKING:
+    from watchdog.observers.api import BaseObserver
+
+log = logging.getLogger("questionpy-sdk:watcher")
+
+_DEBOUNCE_INTERVAL = 0.5  # seconds
+
+
+class _EventHandler(FileSystemEventHandler):
+    """Debounces events for watchdog file monitoring, ignoring events in the `dist` directory."""
+
+    def __init__(
+        self, loop: asyncio.AbstractEventLoop, notify_callback: Callable[[], Awaitable[None]], watch_path: Path
+    ) -> None:
+        self._loop = loop
+        self._notify_callback = notify_callback
+        self._watch_path = watch_path
+
+        self._event_debouncer = EventDebouncer(_DEBOUNCE_INTERVAL, self._on_file_changes)
+
+    def start(self) -> None:
+        self._event_debouncer.start()
+
+    def stop(self) -> None:
+        if self._event_debouncer.is_alive():
+            self._event_debouncer.stop()
+            self._event_debouncer.join()
+
+    def dispatch(self, event: FileSystemEvent) -> None:
+        # filter events and debounce
+        if not self._ignore_event(event):
+            self._event_debouncer.handle_event(event)
+
+    def _on_file_changes(self, events: list[FileSystemEvent]) -> None:
+        # skip synchronization hassle by delegating this to the event loop in the main thread
+        asyncio.run_coroutine_threadsafe(self._notify_callback(), self._loop)
+
+    def _ignore_event(self, event: FileSystemEvent) -> bool:
+        """Ignores events that should not trigger a rebuild.
+
+        Args:
+            event: The event to check.
+
+        Returns:
+            `True` if event should be ignored, otherwise `False`.
+        """
+        if isinstance(event, FileOpenedEvent | FileClosedEvent):
+            return True
+
+        # ignore events events in `dist` dir
+        relevant_path = event.dest_path if isinstance(event, FileSystemMovedEvent) else event.src_path
```

Comment:
Beim Verschieben wird doch der src_path gelöscht. Könnte das nicht auch relevant sein? Also wenn ich etwas aus dem source Verzeichnis in das dist Verzeichnis verschiebe.

Reply:
Ja, richtig. Aber dem Fall gibt es ein separates `modified`-Event für das source-Verzeichnis, welches nicht ignoriert wird.